core/math/ease/ease.odin

```diff
@@ -0,0 +1,466 @@
+
+		case .Elastic_In: return elastic_in(p)
+		case .Elastic_Out: return elastic_out(p)
+		case .Elastic_In_Out: return elastic_in_out(p)
+
+		case .Back_In: return back_in(p)
+		case .Back_Out: return back_out(p)
+		case .Back_In_Out: return back_in_out(p)
+
+		case .Bounce_In: return bounce_in(p)
+		case .Bounce_Out: return bounce_out(p)
+		case .Bounce_In_Out: return bounce_in_out(p)
+	}
+
+	// in case type was invalid
+	return 0
+}
+
+Flux_Map :: struct($T: typeid) {
+	values: map[^T]Flux_Tween(T),
+}
+
+Flux_Tween :: struct($T: typeid) {
+	value: ^T,
+	start: T,
+	diff: T,
+	goal: T,
+
+	// using ticks for timing instead
+	delay_tick_start: time.Tick,
+	delay: time.Duration,
+	duration: time.Duration,
+
+	progress: f64,
+	rate: f64,
+	type: Ease,
+
+	inited: bool,
+
+	// callbacks, data can be set, will be pushed to callback
+	data: rawptr, // by default gets set to value input
+	on_start: proc(flux: ^Flux_Map(T), data: rawptr),
+	on_update: proc(flux: ^Flux_Map(T), data: rawptr),
+	on_complete: proc(flux: ^Flux_Map(T), data: rawptr),
+}
+
+// init flux map to a float type and a wanted cap
+flux_init :: proc($T: typeid, cap := 8) -> Flux_Map(T) where intrinsics.type_is_float(T) {
+	return {
+		make(map[^T]Flux_Tween(T), cap),
+	}
+}
+
+// delete map content
+flux_destroy :: proc(flux: Flux_Map($T)) where intrinsics.type_is_float(T) {
+	delete(flux.values)
+}
+
+// clear map content, stops all animations
+flux_clear :: proc(flux: ^Flux_Map($T)) where intrinsics.type_is_float(T) {
+	clear(&flux.values)
+}
+
+// append / overwrite existing tween value to parameters
+// rest is initialized in flux_tween_init, inside update
+// return value can be used to set callbacks
+flux_to :: proc(
+	flux: ^Flux_Map($T),
+	value: ^f32, 
+	goal: f32, 
+	type: Ease = .Quadratic_Out,
+	duration: time.Duration = time.Second, 
+	delay: time.Duration = 0,
+) -> (tween: ^Flux_Tween(T)) where intrinsics.type_is_float(T) {
+	if res, ok := &flux.values[value]; ok {
+		tween = res
+	} else {
+		flux.values[value] = {}
+		tween = &flux.values[value]
+	}
+
+	tween^ = { 
+		value = value, 
+		goal = goal, 
+		duration = duration,
+		delay = delay,
+		delay_tick_start = time.tick_now(),
+		type = type,
+		data = value,
+	}
+
+	return
+}
+
+// init internal properties
+flux_tween_init :: proc(tween: ^Flux_Tween($T), duration: time.Duration) where intrinsics.type_is_float(T) {
+	tween.inited = true
+	tween.start = tween.value^
+	tween.diff = tween.goal - tween.value^
+	s := time.duration_seconds(duration)
+	tween.rate = duration > 0 ? 1.0 / s : 0
+	tween.progress = duration > 0 ? 0 : 1
+}
+
+// update all tweens, wait for their delay if one exists
+// calls callbacks in all stages, when they're filled
+// deletes tween from the map after completion
+flux_update :: proc(flux: ^Flux_Map($T), dt: f64) where intrinsics.type_is_float(T) {
+	size := len(flux.values)
+	now := time.tick_now()
+
+	for key, tween in &flux.values {
+		if tween.delay != 0 {
```

Comment:
Is this correct? The way I read it, if you make a tween with a delay of 1 second and update it each frame, then any remainder of the delay is lost.

i.e.: a tween with a delay of 1 second, updated at a `dt` of 0.45:
0.45
0.9
1.35 ... the 0.35 remainder gets ignored instead of forwarding the tween by that much?

Reply:
Also, why `now` and `diff` to figure out how much the delay has advanced by instead of subtracting `dt` from `delay` until `delay <= 0` and using any remainder as the rest of `dt` to then advance the tween by?

Reply:
> 

thats how it works in the original flux repo https://github.com/rxi/flux/blob/master/flux.lua#L150
cant remember why i changed it but i think the using core time was more correct? gonna try it out again

Reply:
I see. I just think it's going to be surprising if you have frame stutters and your delay ends up longer than specified.

At 60fps any `dt` remainder in the `delay` being ignored won't be obvious, but if `dt` is big then a (sub)second delay before the animation starts will effectively be stretched until the next update, and that'll be noticeable.

Reply:
> Is this correct? The way I read it, if you make a tween with a delay of 1 second and update it each frame, then any remainder of the delay is lost.
> 
> i.e.: a tween with a delay of 1 second, updated at a `dt` of 0.45: 0.45 0.9 1.35 ... the 0.35 remainder gets ignored instead of forwarding the tween by that much?

using the core:time tick diffs arent based on the dt at all, so i dont get what you mean :sweat_smile:

Reply:
That's exactly what I mean. It keeps track of the delay separately to `dt`, which is just weird. It means that if you don't update it for a second and only then start pumping the update again, the delay will have already passed.

Or if you update it once every 450ms for some reason (doing a lot of processing), then instead of 350ms being applied to your tween, they get dropped on the floor because they were left over from the delay.

Why is `delay` decoupled from `dt` instead of being advanced by it like the tween itself is? It's strange and leads to unexpected situations when you have long frames.

Reply:
that makes sense yeah, i'll try out the `dt` based approach again :+1: